### ISSUE TITLE
Full Recalculation Grandparent Rollup not calculating SUM correctly of all child records

### ISSUE THREAD

**Issue by baobao917** [jamessimone/apex-rollup#523]
Hi @jamessimone - Opening a new issue, as discussed, separate from #500 .

After implementing v.1.6.0, I tested the full recalculation from the Recalculate Rollup tab on a grandparent rollup.  I then ran a report summing the grandchild record amounts and compared them against the grandparent actual rollup calculated amount.  Some did calculate correctly, but a fairly large number did not.  See below the details:

Video - https://www.loom.com/share/b12ea6de638241c6a4a1a61d441ff2a9
From video - Exported report from production (excel) - 
[Asset Plan Spend Grandparent Rollup QA calculation.xlsx](https://github.com/jamessimone/apex-rollup/files/13049175/Asset.Plan.Spend.Grandparent.Rollup.QA.calculation.xlsx)

I believe you have access to the sandbox org still.  I created the same report in sandbox (as shown in video). Link is below
https://serviceone--uat.sandbox.lightning.force.com/lightning/r/Report/00O7e000001A9HhEAK/view?queryScope=userFolders

**Comment by jamessimone** [comment]
@baobao917 finally had the chance to break ground on this issue today. Many thanks, again, for your videos and writeups. As mentioned on #500, grandparent full recalcs work slightly differently from standard rollups, which is why I thought it best to break this out into a separate issue. I'm still hopeful to include this fix in 1.6.1 - I will keep you updated as to my progress.

**Comment by baobao917** [comment]
Thanks @jamessimone .  I'm going to have to re-open #500 as well.  I went thru an exercise (similar as to the one here for grandparent rollup full recalculations) to validate the parent rollup full recalculations for all of the parent records.

**Comment by jamessimone** [comment]
@baobao917 I started looking into this just now, starting with [this Asset record](https://serviceone--uat.sandbox.lightning.force.com/lightning/r/Asset/02i4w00000Z47H6AAJ/view). By default, Apex Rollup lets updates silently fail when there's an issue updating a parent record. This is colloquially described as the "all or none" database setting, where a `true` value shows exceptions on save and a `false` value (what Apex Rollup uses) leads to this silent failure. If I temporarily update the default value for "all or none" to `true`, here's what I see:

```
FATAL_ERROR System.DmlException: Update failed. First exception on row 0 with id 02i4w00000Z47H6AAJ; 
first error: REQUIRED_FIELD_MISSING, Required fields are missing: [Equipment Type]: [Equipment Type]
```

I get the same error with the second row in your report as well. Any chance the Assets that weren't properly updated are all missing this required value?

**Comment by baobao917** [comment]
Hi James - I can def check that and will let you know. Where is this
setting "all or none"? Will it throw an error if I set this setting (and if
so, how? Via email?)

Also if I understand correctly, then the asset record shouldn't have had
any roll-up calculation populated into the roll-up field correct? (Since
that exception would have prevented the parent record from saving?)

Bc I was definitely seeing some records where the roll-up field was
populated but with an incorrect value.

**Comment by jamessimone** [comment]
@baobao917 at the moment "all or none" isn't an exposed setting within Apex Rollup. I could change that in the future.

Yes, your understanding is correct - Apex Rollup would only have been able to update records where the parent record can be saved successfully. It could just be that the first two records I checked happened to be missing that field and that there were legitimate errors lower down on the report - happy to give that one a look over once you've updated the assets!

**Comment by baobao917** [comment]
@jamessimone Yes it would be nice if we could know if something was preventing the rollup from calculating correctly.  

I probably won't be able to update the Asset record until this Friday or next week.

**Comment by jamessimone** [comment]
Interesting. I haven't had the chance to take a look yet. I am planning to come back to this one - going to wrap up the next release with the full fix for #518 tomorrow. The only thing I can think of, at the moment, is trying the full recalc again with a smaller batch size to see if there were any timeout issues causing some of the grandparent records to not be updated. It could also be that there are other validation errors preventing the Assets from saving, as was the case in the sandbox. I won't be able to get back to investigating this until the week of November 6th.

**Comment by baobao917** [comment]
@jamessimone Are you referring to the batch chunk size in the Rollup Control?  Suggestions as to what size I should try?

![image](https://github.com/jamessimone/apex-rollup/assets/104730171/0b7e7658-0a2d-4f09-8687-0ddd5a713f68)

**Comment by jamessimone** [comment]
Yes, I'd try something lower, like 200, and see if that makes a difference. It'll be a good data point, either way.

**Comment by baobao917** [comment]
@jamessimone I tried a batch chunk size = 200.  Actually have more exceptions now than before (3,538 vs 3,020)...but that could be due to more data as this is a production org.  

Thanks for the input....I'll look forward to when you are able to get to this.  Thanks!

**Comment by jamessimone** [comment]
@baobao917 that's curious, the batch size shouldn't change anything other than throttling the overall number of parent records getting updated - which, in an automation-heavy org, would give you more breathing room. It doesn't affect the logic for how rollups are calculated. I suppose that if I had access to the org, I could at least get into the logs and examine a few of the parent records not properly updating to get a better understanding of what the issue is.

**Comment by baobao917** [comment]
@jamessimone I will work on getting you access to that org.  Will be in touch

**Comment by baobao917** [comment]
@jamessimone I will be providing you your own user login for this org.  You should be getting an email shortly.  Just pls remember that this is a production org.  Thanks!

**Comment by jamessimone** [comment]
@baobao917 I can confirm I got in. I will be reviewing the report you linked previously and running a full recalc to check logs tomorrow

**Comment by jamessimone** [comment]
@baobao917 - I ran a few recalcs today. As far as I can tell, this is once again an issue with some sort of unrelated DML exception on the parent records. I never saw evidence of a parent record not having the right value get calculated; some of the parents clearly had been updated at some point in the past with a previously correct value, but that now-stale value hasn't been able to be updated by any of the subsequent full recalcs due to whatever silent DML failure is occurring. I am going to be adding some logging in the next version that will, when logging is enabled, create `ERROR` logging level entries that you can examine after the fact to get insight into the nature of what is preventing each parent record from being updated.

**Comment by baobao917** [comment]
I thought it was strange that if the grandparent record had some DML
exception that silently failed and prevented it from getting updated, then
I would have expected it to be a $0 value.  However the fact it had any
value at all meant it was able to be updated during the recalc. For
context, prior to running the full recalc all of the grandparent target
roll-up field had a null value bc this is a brand new field which would t
have had any value populated prior to the ApexRollup full recalc job.

Does that make sense?

Thanks,
Bao

**Comment by jamessimone** [comment]
That does make sense, but could there have been other validation rule or new field changes going on during the initial full recalc operation? That's the only explanation I can offer; as far as what I saw, the logs contained the right values (the ones that you had in your spreadsheet), but once v1.6.2 is released we'll also be able to see what - if any - is being reported by the database for record update failures

**Comment by baobao917** [comment]
@jamessimone I've upgraded production to v1.6.2.  Pls let me know if you need anything from me.  Thanks!

**Comment by jamessimone** [comment]
@baobao917 I can run a full recalc there tomorrow if you'd like. I will just be looking to go through the logs generated by those runs to see which parent-level records are having DML issues with updating.

**Comment by baobao917** [comment]
Yes please feel free to run a full recalc on that particular roll up. No
one is using that data yet bc I've told the client issues are still being
worked out.

Just not on the other roll ups pls. Got those squared away.  Thanks!

**Comment by jamessimone** [comment]
@baobao917 I was able to find the source of the issue while running the full recalc this morning. I'll include a fix for this in the next version

**Comment by baobao917** [comment]
Awesome, thanks!

Bao Do, CPA

Principal

*Accretive Advisory LLC* Annapolis, MD 21403

410-541-6082 | ***@***.***

/in/bdo0917 <https://www.linkedin.com/in/bdo0917/>

https://trailblazer.me/id/baodo

**Comment by baobao917** [comment]
@jamessimone I tested this full recalc issue on grandparent rollups against v1.6.3.  It significantly improved the problem but there were still a small number of records where it still didn't calculate.  I recorded a video and the two excel file calculations I referenced in the video.

Video - https://www.loom.com/share/1dbe7a8bb62447adb69e09e488e281c5

Excel File (after 1st full recalc) - 
[Asset Plan Spend Grandparent Rollup QA calculation-PROD v3.xlsx](https://github.com/jamessimone/apex-rollup/files/13401715/Asset.Plan.Spend.Grandparent.Rollup.QA.calculation-PROD.v3.xlsx)

Excel File (after 2nd full recalc) - 
[Asset Plan Spend Grandparent Rollup QA calculation-PROD v4.xlsx](https://github.com/jamessimone/apex-rollup/files/13401716/Asset.Plan.Spend.Grandparent.Rollup.QA.calculation-PROD.v4.xlsx)
